libs/xpdf/xpdf-src/doc/pdftohtml.1
.\" Copyright 1997-2025 Glyph & Cog, LLC
.TH pdftohtml 1 "14 Nov 2025"
.SH NAME
pdftohtml \- Portable Document Format (PDF) to HTML converter
(version 4.06)
.SH SYNOPSIS
.B pdftohtml
[options]
.I PDF-file
.I HTML-dir
.SH DESCRIPTION
.B Pdftohtml
converts Portable Document Format (PDF) files to HTML.
.PP
Pdftohtml reads the PDF file,
.IR PDF-file ,
and places an HTML file for each page, along with auxiliary images
in the directory,
.IR HTML-dir .
The HTML directory will be created; if it already exists, pdftohtml
will report an error.
.SH CONFIGURATION FILE
Pdftohtml reads a configuration file at startup.  It first tries to
find the user's private config file, ~/.xpdfrc.  If that doesn't
exist, it looks for a system-wide config file, typically /etc/xpdfrc
(but this location can be changed when pdftohtml is built).  See the
.BR xpdfrc (5)
man page for details.
.SH OPTIONS
Many of the following options can be set with configuration file
commands.  These are listed in square brackets with the description of
the corresponding command line option.
.TP
.BI \-f " number"
Specifies the first page to convert.
.TP
.BI \-l " number"
Specifies the last page to convert.
.TP
.BI \-z " number"
Specifies the initial zoom level.  The default is 1.0, which means
72dpi, i.e., 1 point in the PDF file will be 1 pixel in the HTML.
Using \'-z 1.5', for example, will make the initial view 50% larger.
.TP
.BI \-r " number"
Specifies the resolution, in DPI, for background images.  This
controls the pixel size of the background image files.  The initial
zoom level is controlled by the \'-z' option.  Specifying a larger
\'-r' value will allow the viewer to zoom in farther without upscaling
artifacts in the background.
.TP
.B \-noframe
The default is to generate an \'index.html\' page with a sidebar
listing all of the pages and a viewing area containing the current
page in an HTML iframe. This option switches back to the older
(pre-4.06) behavior of generating a very simple index page that just
has links to the pages.
.TP
.BI \-vstretch " number"
Specifies a vertical stretch factor.  Setting this to a value greater
than 1.0 will stretch each page vertically, spreading out the lines.
This also stretches the background image to match.
.TP
.B \-embedbackground
Embeds the background image as base64-encoded data directly in the
HTML file, rather than storing it as a separate file.
.TP
.B \-nofonts
Disable extraction of embedded fonts.  By default, pdftohtml extracts
TrueType and OpenType fonts.  Disabling extraction can work around
problems with buggy fonts.
.TP
.B \-embedfonts
Embeds any extracted fonts as base64-encoded data directly in the HTML
file, rather than storing them as separate files.
.TP
.B \-skipinvisible
Don't draw invisible text.  By default, invisible text (commonly used
in OCR'ed PDF files) is drawn as transparent (alpha=0) HTML text.
This option tells pdftohtml to discard invisible text entirely.
.TP
.B \-allinvisible
Treat all text as invisible.  By default, regular (non-invisible) text
is not drawn in the background image, and is instead drawn with HTML
on top of the image.  This option tells pdftohtml to include the
regular text in the background image, and then draw it as transparent
(alpha=0) HTML text.
.TP
.B \-formfields
Convert AcroForm text and checkbox fields to HTML input elements.
This also removes text (e.g., underscore characters) and erases
background image content (e.g., lines or boxes) in the field areas.
.TP
.B \-meta
Include PDF document metadata as 'meta' elements in the HTML header.
.TP
.B \-table
Use table mode when performing the underlying text extraction.  This
will generally produce better output when the PDF content is a
full-page table.  NB: This does not generate HTML tables; it just
changes the way text is split up.
.TP
.B \-overwrite
By default pdftohtml will not overwrite the specified output
directory.  If the directory already exists, pdftohtml will exit with
an error.  This option tells pdftohtml to instead overwrite the
existing directory.
.TP
.BI \-opw " password"
Specify the owner password for the PDF file.  Providing this will
bypass all security restrictions.
.TP
.BI \-upw " password"
Specify the user password for the PDF file.
.TP
.B \-verbose
Print a status message (to stdout) before processing each page.
.RB "[config file: " printStatusInfo ]
.TP
.B \-q
Don't print any messages or errors.
.RB "[config file: " errQuiet ]
.TP
.BI \-cfg " config-file"
Read
.I config-file
in place of ~/.xpdfrc or the system-wide config file.
.TP
.B \-v
Print copyright and version information.
.TP
.B \-h
Print usage information.
.RB ( \-help
and
.B \-\-help
are equivalent.)
.SH BUGS
Some PDF files contain fonts whose encodings have been mangled beyond
recognition.  There is no way (short of OCR) to extract text from
these files.
.SH EXIT CODES
The Xpdf tools use the following exit codes:
.TP
0
No error.
.TP
1
Error opening a PDF file.
.TP
2
Error opening an output file.
.TP
3
Error related to PDF permissions.
.TP
98
Out of memory.
.TP
99
Other error.
.SH AUTHOR
The pdftohtml software and documentation are copyright 1996-2025 Glyph
& Cog, LLC.
.SH "SEE ALSO"
.BR xpdf (1),
.BR pdftops (1),
.BR pdftotext (1),
.BR pdfinfo (1),
.BR pdffonts (1),
.BR pdfdetach (1),
.BR pdftoppm (1),
.BR pdftopng (1),
.BR pdfimages (1),
.BR xpdfrc (5)
.br
.B http://www.xpdfreader.com/


ChangeLog
2025-10-16  Karl Berry  <karl@freefriends.org>

	* Build (status): avoid using egrep.
	Report from Jan Kasprzak,
	https://tug.org/pipermail/tex-k/2025-October/004217.html.

2025-09-07  Karl Berry  <karl@freefriends.org>

	* Makefile.am (install_target): new variable, install-strip by default.
	(world): use it.

2025-07-22  Karl Berry  <karl@freefriends.org>

	* reautoconf: mention that autoconf-only packages need to be
	explicitly added; would be better to read kpse-pkgs.m4.
	Also reautoconf tree for kpse-setup.m4 changes.

2025-06-26  Karl Berry  <karl@freefriends.org>

	* automake 1.18.1.

2025-04-30  Karl Berry  <karl@freefriends.org>

	* Build: incorporate original CFLAGS, CXXFLAGS, OBJCXXFLAGS values.
	https://tug.org/pipermail/tlbuild/2025q2/005771.html

2025-04-29  Karl Berry  <karl@freefriends.org>

	* Build: use TL_CFLAGS, TL_CXXFLAGS, etc., for TL_BUILD_ENV
	if -g is not given.

2025-03-08  Karl Berry  <karl@tug.org>

	* version.ac: 2026/dev.

2025-03-07  Karl Berry  <karl@tug.org>

	* TL'25 release.

2025-01-12  Karl Berry  <karl@tug.org>

	* Build: also support --debug-more for -g -Og -ggdb3,
	per Luigi, as described on https://tug.org/texlive/build.html.

2024-07-12  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake 1.17.

2024-04-29  Karl Berry  <karl@freefriends.org>

	* Makefile.am (distdir): override default dist generation,
	since we don't try to keep make dist working any more.
	https://tug.org/pipermail/tlbuild/2024q2/005547.html

2023-12-31  Karl Berry  <karl@freefriends.org>

	* reautoconf: add -f / --force option to pass to autoreconf.

2023-12-26  Karl Berry  <karl@freefriends.org>

	* reautoconf: don't bother with environment variables;
	because the autotools have many subprograms, setting PATH
	is what makes sense.

2023-12-23  Karl Berry  <karl@tug.org>

	* reautoconf with autoconf-2.72.

2023-12-21  Karl Berry  <karl@freefriends.org>

	* reautoconf (do_it): abort the run if any command fails.

2023-02-25  Karl Berry  <karl@tug.org>

	* Build: show full uname -a.

2023-02-09  Karl Berry  <karl@freefriends.org>

	* version.ac: 2023, no more /dev.

2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2022-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): let's hope for 2019-04-07.

2018-10-02  Karl Berry  <karl@tug.org>

	* Build: TL_MAKE_VERBOSE=VERBOSE=1 by default,
	so that test-suite.log will be shown on failure.
	https://lists.gnu.org/archive/html/automake/2018-09/msg00015.html

2018-04-28  Karl Berry  <karl@tug.org>

	* version.ac: 2019/dev.

2018-04-10  Karl Berry  <karl@tug.org>

	* README.*: update from new tlbuild.

2017-11-15  Karl Berry  <karl@tug.org>

	* *.test: use #!/bin/sh -vx. Add 

2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the

libs/xpdf/xpdf-src/xpdf/pdftohtml.cc
//========================================================================
//
// pdftohtml.cc
//
// Copyright 2005 Glyph & Cog, LLC
//
//========================================================================

#include <aconf.h>
#include <stdio.h>
#include <stdlib.h>
#include "gmem.h"
#include "gmempp.h"
#include "parseargs.h"
#include "gfile.h"
#include "GString.h"
#include "GlobalParams.h"
#include "PDFDoc.h"
#include "HTMLGen.h"
#include "Error.h"
#include "ErrorCodes.h"
#include "config.h"

//------------------------------------------------------------------------

static GBool createBasicIndex(char *htmlDir);
static GBool createFrameIndex(char *htmlDir);

//------------------------------------------------------------------------

static int firstPage = 1;
static int lastPage = 0;
static double zoom = 1;
static int resolution = 150;
static GBool noFrame = gFalse;
static double vStretch = 1;
static GBool embedBackground = gFalse;
static GBool noFonts = gFalse;
static GBool embedFonts = gFalse;
static GBool skipInvisible = gFalse;
static GBool allInvisible = gFalse;
static GBool formFields = gFalse;
static GBool includeMetadata = gFalse;
static GBool tableMode = gFalse;
static GBool overwrite = gFalse;
static char ownerPassword[33] = "\001";
static char userPassword[33] = "\001";
static GBool verbose = gFalse;
static GBool quiet = gFalse;
static char cfgFileName[256] = "";
static GBool printVersion = gFalse;
static GBool printHelp = gFalse;

static ArgDesc argDesc[] = {
  {"-f",                argInt,    &firstPage,       0,
   "first page to convert"},
  {"-l",                argInt,    &lastPage,        0,
   "last page to convert"},
  {"-z",                argFP,     &zoom,            0,
   "initial zoom level (1.0 means 72dpi)"},
  {"-r",                argInt,    &resolution,      0,
   "resolution, in DPI (default is 150)"},
  {"-noframe",          argFlag,   &noFrame,         0,
   "generate a basic index page, without an iframe element"},
  {"-vstretch",         argFP,     &vStretch,        0,
   "vertical stretch factor (1.0 means no stretching)"},
  {"-embedbackground",  argFlag,   &embedBackground, 0,
   "embed the background image as base64-encoded data" },
  {"-nofonts",          argFlag,   &noFonts,         0,
   "do not extract embedded fonts"},
  {"-embedfonts",       argFlag,   &embedFonts,      0,
   "embed the fonts as base64-encoded data" },
  {"-skipinvisible",    argFlag,   &skipInvisible,   0,
   "do not draw invisible text"},
  {"-allinvisible",     argFlag,   &allInvisible,    0,
   "treat all text as invisible"},
  {"-formfields",       argFlag,   &formFields,      0,
   "convert form fields to HTML"},
  {"-meta",             argFlag,   &includeMetadata, 0,
   "include document metadata in the HTML output"},
  {"-table",            argFlag,   &tableMode,       0,
   "use table mode for text extraction"},
  {"-overwrite",        argFlag,   &overwrite,       0,
   "overwrite files in an existing output directory"},
  {"-opw",              argString, ownerPassword,    sizeof(ownerPassword),
   "owner password (for encrypted files)"},
  {"-upw",              argString, userPassword,     sizeof(userPassword),
   "user password (for encrypted files)"},
  {"-verbose", argFlag,    &verbose,       0,
   "print per-page status information"},
  {"-q",                argFlag,   &quiet,           0,
   "don't print any messages or errors"},
  {"-cfg",              argString, cfgFileName,      sizeof(cfgFileName),
   "configuration file to use in place of .xpdfrc"},
  {"-v",                argFlag,   &printVersion,    0,
   "print copyright and version info"},
  {"-h",                argFlag,   &printHelp,       0,
   "print usage information"},
  {"-help",             argFlag,   &printHelp,       0,
   "print usage information"},
  {"--help",            argFlag,   &printHelp,       0,
   "print usage information"},
  {"-?",                argFlag,   &printHelp,       0,
   "print usage information"},
  {NULL}
};

//------------------------------------------------------------------------

static int writeToFile(void *file, const char *data, int size) {
  return (int)fwrite(data, 1, size, (FILE *)file);
}

int main(int argc, char *argv[]) {
#if USE_EXCEPTIONS
  try {
#endif

  PDFDoc *doc;
  char *fileName;
  char *htmlDir;
  GString *ownerPW, *userPW;
  HTMLGen *htmlGen;
  GString *htmlFileName, *pngFileName, *pngURL;
  FILE *htmlFile, *pngFile;
  int pg, err, exitCode;
  GBool ok;

  exitCode = 99;

  // parse args
  fixCommandLine(&argc, &argv);
  ok = parseArgs(argDesc, &argc, argv);
  if (printVersion) {
    printf("pdftohtml version %s [www.xpdfreader.com]\n", xpdfVersion);
    printf("%s\n", xpdfCopyright);
    goto err0;
  }
  if (!ok || argc != 3 || printHelp) {
    fprintf(stderr, "pdftohtml version %s [www.xpdfreader.com]\n", xpdfVersion);
    fprintf(stderr, "%s\n", xpdfCopyright);
    printUsage("pdftohtml", "<PDF-file> <html-dir>", argDesc);
    goto err0;
  }
  fileName = argv[1];
  htmlDir = argv[2];

  // read config file
  if (cfgFileName[0] && !pathIsFile(cfgFileName)) {
    error(errConfig, -1, "Config file '{0:s}' doesn't exist or isn't a file",
	  cfgFileName);
  }
  globalParams = new GlobalParams(cfgFileName);
  if (verbose) {
    globalParams->setPrintStatusInfo(verbose);
  }
  if (quiet) {
    globalParams->setErrQuiet(quiet);
  }
  globalParams->setupBaseFonts(NULL);
  globalParams->setTextEncoding("UTF-8");

  // open PDF file
  if (ownerPassword[0] != '\001') {
    ownerPW = new GString(ownerPassword);
  } else {
    ownerPW = NULL;
  }
  if (userPassword[0] != '\001') {
    userPW = new GString(userPassword);
  } else {
    userPW = NULL;
  }
  doc = new PDFDoc(fileName, ownerPW, userPW);
  if (userPW) {
    delete userPW;
  }
  if (ownerPW) {
    delete ownerPW;
  }
  if (!doc->isOk()) {
    exitCode = 1;
    goto err1;
  }

  // check for copy permission
  if (!doc->okToCopy()) {
    error(errNotAllowed, -1,
	  "Copying of text from this document is not allowed.");
    exitCode = 3;
    goto err1;
  }

  // get page range
  if (firstPage < 1) {
    firstPage = 1;
  }
  if (lastPage < 1 || lastPage > doc->getNumPages()) {
    lastPage = doc->getNumPages();
  }

  // create HTML directory
  if (makeDir(htmlDir, 0755)) {
    if (pathIsDir(htmlDir)) {
      if (!overwrite) {
	error(errIO, -1, "HTML output directory '{0:s}' already exists (use '-overwrite' to overwrite it)",
	      htmlDir);
	exitCode = 2;
	goto err1;
      }
    } else {
      error(errIO, -1, "Couldn't create HTML output directory '{0:s}'",
	    htmlDir);
      exitCode = 2;
      goto err1;
    }
  }

  // set up the HTMLGen object
  htmlGen = new HTMLGen(resolution, tableMode);
  if (!htmlGen->isOk()) {
    exitCode = 99;
    goto err1;
  }
  htmlGen->setZoom(zoom);
  htmlGen->setVStretch(vStretch);
  htmlGen->setDrawInvisibleText(!skipInvisible);
  htmlGen->setAllTextInvisible(allInvisible);
  htmlGen->setEmbedBackgroundImage(embedBackground);
  htmlGen->setExtractFontFiles(!noFonts);
  htmlGen->setEmbedFonts(embedFonts);
  htmlGen->setConvertFormFields(formFields);
  htmlGen->setIncludeMetadata(includeMetadata);
  htmlGen->startDoc(doc);

  // convert the pages
  for (pg = firstPage; pg <= lastPage; ++pg) {
    if (globalParams->getPrintStatusInfo()) {
      fflush(stderr);
      printf("[processing page %d]\n", pg);
      fflush(stdout);
    }
    htmlFileName = GString::format("{0:s}/page{1:d}.html", htmlDir, pg);
    pngFileName = GString::format("{0:s}/page{1:d}.png", htmlDir, pg);
    if (!(htmlFile = openFile(htmlFileName->getCString(), "wb"))) {
      error(errIO, -1, "Couldn't open HTML file '{0:t}'", htmlFileName);
      delete htmlFileName;
      delete pngFileName;
      goto err2;
    }
    if (embedBackground) {
      pngFile = NULL;
    } else {
      if (!(pngFile = openFile(pngFileName->getCString(), "wb"))) {
	error(errIO, -1, "Couldn't open PNG file '{0:t}'", pngFileName);
	fclose(htmlFile);
	delete htmlFileName;
	delete pngFileName;
	goto err2;
      }
    }
    pngURL = GString::format("page{0:d}.png", pg);
    err = htmlGen->convertPage(pg, pngURL->getCString(), htmlDir,
			       &writeToFile, htmlFile,
			       &writeToFile, pngFile);
    delete pngURL;
    fclose(htmlFile);
    if (!embedBackground) {
      fclose(pngFile);
    }
    delete htmlFileName;
    delete pngFileName;
    if (err != errNone) {
      error(errIO, -1, "Error converting page {0:d}", pg);
      exitCode = 2;
      goto err2;
    }
  }

  // create the master index
  if (!(noFrame ? createBasicIndex(htmlDir) : createFrameIndex(htmlDir))) {
    exitCode = 2;
    goto err2;
  }

  exitCode = 0;

  // clean up
 err2:
  delete htmlGen;
 err1:
  delete doc;
  delete globalParams;
 err0:

  // check for memory leaks
  Object::memCheck(stderr);
  gMemReport(stderr);

  return exitCode;

#if USE_EXCEPTIONS
  } catch (GMemException e) {
    fprintf(stderr, "Out of memory\n");
    return 98;
  }
#endif
}

static GBool createBasicIndex(char *htmlDir) {
  GString *htmlFileName;
  FILE *html;
  int pg;

  htmlFileName = GString::format("{0:s}/index.html", htmlDir);
  html = openFile(htmlFileName->getCString(), "w");
  if (!html) {
    error(errIO, -1, "Couldn't open HTML file '{0:t}'", htmlFileName);
    delete htmlFileName;
    return gFalse;
  }
  delete htmlFileName;

  fprintf(html, "<html>\n");
  fprintf(html, "<body>\n");
  for (pg = firstPage; pg <= lastPage; ++pg) {
    fprintf(html, "<a href=\"page%d.html\">page %d</a><br>\n", pg, pg);
  }
  fprintf(html, "</body>\n");
  fprintf(html, "</html>\n");

  fclose(html);

  return gTrue;
}

static const char *frameHead =
  "<html>\n"
  "<head>\n"
  "<style type=\"text/css\">\n"
  "body {\n"
  "  height: 100vh;\n"
  "  margin: 0;\n"
  "  border: 0;\n"
  "}\n"
  ".container {\n"
  "  display: flex;\n"
  "  width: 100%;\n"
  "  height: 100%;\n"
  "}\n"
  ".sidebar {\n"
  "  width: 5em;\n"
  "  background-color: #eeeeee;\n"
  "  padding: 0.5em;\n"
  "  overflow: auto;\n"
  "}\n"
  ".sidebar ul {\n"
  "  list-style-type: none;\n"
  "  padding: 0;\n"
  "  margin: 0;\n"
  "}\n"
  ".viewer {\n"
  "  flex: 1;\n"
  "  padding: 0;\n"
  "  overflow: auto;\n"
  "}\n"
  ".viewer iframe {\n"
  "  width: 100%;\n"
  "  height: 100%;\n"
  "  border: none;\n"
  "}\n"
  "</style>\n"
  "</head>\n"
  "<body>\n"
  "<div class=\"container\">\n"
  "  <nav class=\"sidebar\">\n"
  "    <ul>\n";

static const char *frameTail =
  "  </ul>\n"
  "  </nav>\n"
  "  <main class=\"viewer\">\n"
  "    <iframe id=\"viewerFrame\" src=\"page1.html\"></iframe>\n"
  "  </main>\n"
  "</div>\n"
  "<script>\n"
  "  const iframe = document.getElementById(\"viewerFrame\");\n"
  "  document.querySelectorAll('a[page]').forEach(link => {\n"
  "    link.addEventListener('click', (e) => {\n"
  "      e.preventDefault();\n"
  "      const page = link.getAttribute(\"page\");\n"
  "      iframe.src = page;\n"
  "    });\n"
  "  });\n"
  "</script>\n"
  "</body>\n"
  "</html>\n";

static GBool createFrameIndex(char *htmlDir) {
  GString *htmlFileName;
  FILE *html;
  int pg;

  htmlFileName = GString::format("{0:s}/index.html", htmlDir);
  html = openFile(htmlFileName->getCString(), "w");
  if (!html) {
    error(errIO, -1, "Couldn't open HTML file '{0:t}'", htmlFileName);
    delete htmlFileName;
    return gFalse;
  }
  delete htmlFileName;

  fputs(frameHead, html);
  for (pg = firstPage; pg <= lastPage; ++pg) {
    fprintf(html, "      <li><a href=\"#\" page=\"page%d.html\">page %d</a></li>\n",
	    pg, pg);
  }
  fputs(frameTail, html);

  fclose(html);

  return gTrue;
}
